clang/lib/Sema/HLSLExternalSemaSource.cpp

```diff
@@ -323,30 +324,99 @@ struct TemplateParameterListBuilder {
         S.Context, Builder.Record->getDeclContext(), SourceLocation(),
         SourceLocation(), /* TemplateDepth */ 0, Position,
         &S.Context.Idents.get(Name, tok::TokenKind::identifier),
-        /* Typename */ false,
-        /* ParameterPack */ false);
+        /* Typename */ true,
+        /* ParameterPack */ false,
+        /* HasTypeConstraint*/ false);
     if (!DefaultValue.isNull())
       Decl->setDefaultArgument(
           S.Context, S.getTrivialTemplateArgumentLoc(DefaultValue, QualType(),
                                                      SourceLocation()));
-
+    Decl->setReferenced();
```

Comment:
Is this a good bugfix we'd want independently of the concepts change? Or is this something that had to change because we're also using concepts now?

Reply:
This is an odd addition that I am not 100% convinced is absolutely necessary, although I haven't tried testing without it. I don't think it's a bugfix per se. The reason this line exists is because I am trying to match this AST as perfectly as possible, and it served as a guide throughout my development (Thanks to Chris):
https://godbolt.org/z/Gq4hdTcrK
The AST that is constructed in this PR is constructed *as if* the code in the godbolt link were defined and parsed somewhere, and the code in the link uses typename and is referenced. So, I set the appropriate decl as referenced, and make sure it's represented with a typename rather than a class.

Reply:
@llvm-beanz  / @bogner - did you write this originally?  Do you have any insights on what the right thing to do here is?

Reply:
`referenced' means the declaration is referenced/used somewhere in the source code, you should not be setting it here.